`rubbl_casatables_impl` does not play nicely with dynamically linked casacore

So the final hurdle I need to overcome is integrate rubbl into a library which depends on [aoflagger](https://gitlab.com/aroffringa/aoflagger), (a C++ library), which in turn dynamically links to system casacore. 

[Here's an example of what happens when you try to do this](https://github.com/derwentx/rubbl/tree/aoflagger)

```bash
cargo run --example aoflagger
```

```txt
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.16gq24j9fog1euvt.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.17pa5xmv3az8hq3g.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.1f5yonli7ukejw7a.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.1kgv2n5r190gpnsw.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.1lkxqrjba2xfjqsk.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.201eou91xu6cxmfz.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.23iwc8ya460o7p5i.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.247u5mhsjp75cyle.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.2dfmb8s85fgf31bv.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.2e2l0ukqy9nx9ehl.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.2e6qub8bm1vg13p.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.2joziq0hls5bh2gi.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.2otdyesvk4me15nh.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.321wnz0hbozpfigt.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.32iiy9l33y4g2swk.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.3900oz659edid5qo.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.3a96ute0ls60a53m.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.3euxygl2avtvn365.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.3h5uu0awzm8sulwv.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.42sc9o1k9vtv9j5k.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.45o41tn7z6499bf0.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.46rubjlz441yk1jp.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4bwms4rodrcnguac.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4cvk50yxv7r1qn5s.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4g5rm1s91v7kigpi.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4jewhwjrwmaobcnr.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4msmtyriae1pdgak.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4v7ix4w1wxtl300y.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4vqeqn47p0mi9rz0.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4y3yiqdl33juj4i2.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4zpvze6d4a4c2z1a.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.526h85metrm87txt.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.5bpvcll37byiteua.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.8tscr8hn6nrc7vr.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.cf7mal42n678zuk.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.fhg5faix96dn6l3.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.mfabumz5pdiev6o.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.p1v1tkud1s94m6w.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.t24lxj1let3es2x.rcgu.o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9.4l7n8ep4eukhnoww.rcgu.o" "-Wl,--as-needed" "-L" "/home/ubuntu/rubbl/target/debug/deps" "-L" "/home/ubuntu/rubbl/target/debug/build/rubbl_casatables-03634f81c56c59eb/out" "-L" "/home/ubuntu/rubbl/target/debug/build/aoflagger_sys-eaa1edc3a138f823/out" "-L" "/home/ubuntu/rubbl/target/debug/build/cxx-7bcb08d8cdc2b483/out" "-L" "/home/ubuntu/rubbl/target/debug/build/link-cplusplus-c65b8acadc1dd30e/out" "-L" "/home/ubuntu/rubbl/target/debug/build/rubbl_casatables_impl-259c4b3eb238d806/out" "-L" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/ubuntu/rubbl/target/debug/deps/libaoflagger_sys-17a7e2441d33c252.rlib" "/home/ubuntu/rubbl/target/debug/deps/libcxx-fa5e1c8b53f53b74.rlib" "/home/ubuntu/rubbl/target/debug/deps/liblink_cplusplus-de8ee052f8c1a6d2.rlib" "/home/ubuntu/rubbl/target/debug/deps/librubbl_casatables-877bb02c726e0b5f.rlib" "/home/ubuntu/rubbl/target/debug/deps/librubbl_core-392e99bbdee78901.rlib" "/home/ubuntu/rubbl/target/debug/deps/libtermcolor-8f27c150af643511.rlib" "/home/ubuntu/rubbl/target/debug/deps/libbyteorder-f387c8e731df044c.rlib" "/home/ubuntu/rubbl/target/debug/deps/libapprox-dbdf745f86c8cf83.rlib" "/home/ubuntu/rubbl/target/debug/deps/libndarray-143b8cd2a5116d4a.rlib" "/home/ubuntu/rubbl/target/debug/deps/libmatrixmultiply-d39b18112f72a750.rlib" "/home/ubuntu/rubbl/target/debug/deps/libnum_complex-9acdff459ef95ba4.rlib" "/home/ubuntu/rubbl/target/debug/deps/libnum_integer-9f8a31d45bf1668f.rlib" "/home/ubuntu/rubbl/target/debug/deps/libnum_traits-fb574b6b6a0d3398.rlib" "/home/ubuntu/rubbl/target/debug/deps/librawpointer-6aa2ca10ed178081.rlib" "/home/ubuntu/rubbl/target/debug/deps/libfailure-fdb1d3c84a472728.rlib" "/home/ubuntu/rubbl/target/debug/deps/libbacktrace-ce06b6f72e2d0b71.rlib" "/home/ubuntu/rubbl/target/debug/deps/libminiz_oxide-c96737ccceaf4411.rlib" "/home/ubuntu/rubbl/target/debug/deps/libadler-005892d84f32fadf.rlib" "/home/ubuntu/rubbl/target/debug/deps/libobject-e8ea05f7d19f2c4e.rlib" "/home/ubuntu/rubbl/target/debug/deps/libaddr2line-cf893eb4714a5533.rlib" "/home/ubuntu/rubbl/target/debug/deps/libgimli-96bbd316763549a0.rlib" "/home/ubuntu/rubbl/target/debug/deps/libcfg_if-dece450273084492.rlib" "/home/ubuntu/rubbl/target/debug/deps/librustc_demangle-0cc08d6079258aff.rlib" "/home/ubuntu/rubbl/target/debug/deps/libclap-b1a1af300c325eed.rlib" "/home/ubuntu/rubbl/target/debug/deps/libvec_map-5f104c2ee05f26e6.rlib" "/home/ubuntu/rubbl/target/debug/deps/libtextwrap-349506f3b2349b3f.rlib" "/home/ubuntu/rubbl/target/debug/deps/libunicode_width-05a4b639dec6224c.rlib" "/home/ubuntu/rubbl/target/debug/deps/libstrsim-0fc606efc91e3789.rlib" "/home/ubuntu/rubbl/target/debug/deps/libbitflags-117d41ac256572da.rlib" "/home/ubuntu/rubbl/target/debug/deps/libatty-bc49c46593d1386b.rlib" "/home/ubuntu/rubbl/target/debug/deps/liblibc-c66dbc8ff67d320c.rlib" "/home/ubuntu/rubbl/target/debug/deps/libansi_term-329dab68dba20daf.rlib" "-Wl,--start-group" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-008055cc7d873802.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-06f01ac2578bda94.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-f9a3c3274a1835e0.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-d4cbb754ee9f4daa.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-95c14e1c1e3ebcc4.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-d489f0ca872880cc.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-75f07df0b18fea39.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-0c35b278736219a2.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-e530649c9a06e3c6.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-6b148909d375a785.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-cd15fa647f4775d1.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-74be3a703f788ba2.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-8f2c5b445c28b2e3.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-8480e85e0be96197.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-ac23a75f6f42004e.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-557ba8776e04d182.rlib" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-4beb03d03503c439.rlib" "-Wl,--end-group" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-dd7db1bec6909f24.rlib" "-Wl,-Bdynamic" "-laoflagger" "-lstdc++" "-lstdc++" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/ubuntu/rubbl/target/debug/examples/aoflagger-b459d9881107dbf9" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs"
  = note: /usr/bin/ld: /home/ubuntu/rubbl/target/debug/deps/librubbl_casatables-877bb02c726e0b5f.rlib(Aipsrc.o): undefined reference to symbol '_ZN8casacore6Aipsrc11findNoParseERNS_6StringERKS1_j'
          /usr/bin/ld: /lib/x86_64-linux-gnu/libcasa_casa.so.4: error adding symbols: DSO missing from command line
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

warning: `rubbl_casatables` (example "aoflagger") generated 2 warnings
error: could not compile `rubbl_casatables` due to previous error; 2 warnings emitted
```

I was wondering if you had any suggestions on how to approach this problem. I'm thinking that adding a feature which statically compiles rubbl_casatables_impl into rubbl_casatables would be a good way to do this, but I'm open to suggestions.

Cheers!

--- a/casatables/build.rs
+++ b/casatables/build.rs
@@ -12,6 +12,9 @@ fn main() {
         .cpp(true)
         .warnings(true)
         .flag_if_supported("-std=c++11")
+        // This allows us to treat rubbl's modified casacore as a separate 
+        // namespace, so that both vanilla casacore and rubbl can be linked
+        // at the same time. 
         .define("casacore", "rubbl_casacore")
         .include("src")
         .include(env::var_os("DEP_CASA_INCLUDE").unwrap())

--- a/casatables/build.rs
+++ b/casatables/build.rs
@@ -12,6 +12,7 @@ fn main() {
         .cpp(true)
         .warnings(true)
         .flag_if_supported("-std=c++11")
+        .define("casacore", "rubbl_casacore")
         .include("src")
         .include(env::var_os("DEP_CASA_INCLUDE").unwrap())
         .files(FILES)

--- a/casatables/src/lib.rs
+++ b/casatables/src/lib.rs
@@ -1780,9 +1780,9 @@ mod tests {
     use super::*;
     use crate::glue::{GlueDataType, TableDescCreateMode};
     use ndarray::array;
-    use rubbl_core::Complex;
     use tempfile::tempdir;
 
+    #[allow(non_camel_case_types)]
     type c64 = Complex<f64>;
 
     #[test]

--- a/casatables_impl/build.rs
+++ b/casatables_impl/build.rs
@@ -8,7 +8,11 @@ fn main() {
         .cpp(true)
         .warnings(true)
         .flag_if_supported("-std=c++11")
+        // This allows us to treat rubbl's modified casacore as a separate 
+        // namespace, so that both vanilla casacore and rubbl can be linked
+        // at the same time. 
         .define("casacore", "rubbl_casacore")
+        // Without this, using casa in multiple threads causes segfaults
         .define("USE_THREADS", "1")
         .include(".")
         .files(FILES)

--- a/casatables_impl/build.rs
+++ b/casatables_impl/build.rs
@@ -8,6 +8,7 @@ fn main() {
         .cpp(true)
         .warnings(true)
         .flag_if_supported("-std=c++11")
+        .define("casacore", "rubbl_casacore")
         .define("USE_THREADS", "1")
         .include(".")
         .files(FILES)

